pkg/kubelet/util/manager/watch_based_manager.go

```diff
@@ -105,10 +164,16 @@ func (c *objectCache) newStore() cache.Store {
 }
 
 func (c *objectCache) newReflector(namespace, name string) *objectCacheItem {
+	initialized := false
+
 	fieldSelector := fields.Set{"metadata.name": name}.AsSelector().String()
 	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
 		options.FieldSelector = fieldSelector
-		return c.listObject(namespace, options)
+		object, err := c.listObject(namespace, options)
+		if err == nil {
+			initialized = true
```

Comment:
This isn't correct. The fact that list operation finished doesn't mean that the cache is already initialized.
With this approach you are exposed to race where list operation finished, but the store.Replace() wasn't yet called.

I consciously proposed the approach where you need to override Replace() on your implementation on store to set the "initialized" param after Replace() is finished.

Reply:
Thanks. I got a new understanding of the client-go. I modified the code and added a new Replace function to change initialized flag.

Reply:
The new solution will have DATA_RACE, I tried to fixed it, but failed.
I think we want to get the latest data when rerun the Reflector in the new solution. But in the actual scenario, we will re-create the watch link after 5 resyncInterval cycles. the time is about 5 minutes. after 5 minutes, kube-apiserver B will not be in the old state.
also, if kube-apiserver A and B are in different states for a long time, the pods in different nodes of the same deployment will also use different configmap information. currently, we cannot solve this problem.

Can we use the old solution(just create a new Reflector),  this solution is more simpler.

Reply:
> Can we use the old solution(just create a new Reflector), this solution is more simpler.

No - because that can technically go back in time, which we definitely don't want.

> also, if kube-apiserver A and B are in different states for a long time, the pods in different nodes of the same deployment will also use different configmap information. currently, we cannot solve this problem.

That's a different thing than doing back in time.

> I think we want to get the latest data when rerun the Reflector in the new solution. But in the actual scenario, we will re-create the watch link after 5 resyncInterval cycles. the time is about 5 minutes. after 5 minutes, kube-apiserver B will not be in the old state.

Sorry - I don't undertand.
Calling reflector.Run() will immediately call "List()" which will initialize correctly.

Reply:
I understand. I have fixed the data race by add a lock to wait.exponentialBackoffManagerImpl.